File: julia/ch1/ex106.jl
#=
ex106:
- Julia version: 0.6.4
- Author: Yohan
- Date: 2018-08-01
=#

function square(x)
  x * x
end

function average(x, y)
  (x + y) / 2.0
end

function improve(guess, x)
  average(guess, x / guess)
end

function good_enough(guess, x)
  abs((square(guess) - x)) < 0.001
end

function sqrt_iter(guess, x)
  good_enough(guess, x) ? guess : sqrt_iter(improve(guess, x), x)
end

function my_sqrt(x)
  sqrt_iter(1.0, x)
end

function bad_sqrt_iter(guess, x)
  new_if(good_enough(guess, x), guess, bad_sqrt_iter(improve(guess, x), x))
end

function bad_sqrt(x)
  bad_sqrt_iter(1.0, x)
end

function new_if(predicate, then_clause, else_clause)
  predicate ? then_clause : else_clause
end

File: elixir/ch1/ex106.ex
# import_file("Ex106.ex")

defmodule Ex106 do
  @moduledoc false

  def square(x), do: x * x

  def average(x, y), do: (x + y) / 2.0

  def improve(guess, x), do: average(guess, x / guess)

  def good_enough(guess, x), do: abs(square(guess) - x) < 0.001

  def sqrt_iter(guess, x) do
    cond do
      good_enough(guess, x) -> guess
      true -> sqrt_iter(improve(guess, x), x)
    end
  end

  def sqrt(x), do: sqrt_iter(1.0, x)

  def new_if(true, then_clause, _), do: then_clause
  def new_if(false, _, else_clause), do: else_clause

  def bad_sqrt_iter(guess, x), do: new_if(good_enough(guess, x), guess, bad_sqrt_iter(improve(guess, x), x))

  # will this hang? Yes
  def bad_sqrt(x), do: bad_sqrt_iter(1.0, x)
end
